src/confidential/elip151.rs

```diff
@@ -0,0 +1,207 @@
+//! https://github.com/ElementsProject/ELIPs/blob/main/elip-0151.md
+//!
+
+use bitcoin::hashes::{sha256t_hash_newtype, Hash};
+use bitcoin::secp256k1;
+use bitcoin::Network;
+use elements::encode::Encodable;
+use elements::opcodes;
+use elements::script::Builder;
+
+use crate::confidential::{Descriptor as ConfidentialDescriptor, Key};
+use crate::descriptor::{DescriptorSecretKey, SinglePriv};
+use crate::extensions::{Extension, ParseableExt};
+use crate::{Descriptor as OrdinaryDescriptor, DescriptorPublicKey, Error};
+
+/// The SHA-256 initial midstate value for the [`Elip151Hash`].
+const MIDSTATE_ELIP151: [u8; 32] = [
+    0x49, 0x81, 0x61, 0xd8, 0x52, 0x45, 0xf7, 0xaa, 0xd8, 0x24, 0x27, 0xb5, 0x64, 0x69, 0xe7, 0xd6,
+    0x98, 0x17, 0xeb, 0x0f, 0x27, 0x14, 0x6f, 0x4e, 0x7b, 0x95, 0xb3, 0x6e, 0x46, 0xc1, 0xb5, 0x61,
+];
+
+sha256t_hash_newtype!(
+    Elip151Hash,
+    Elip151Tag,
+    MIDSTATE_ELIP151,
+    64,
+    doc = "ELIP-151 Deterministic descriptor blinding keys",
+    forward
+);
+
+impl Key {
+    pub fn from_elip151<T: Extension + ParseableExt>(
+        descriptor: &OrdinaryDescriptor<DescriptorPublicKey, T>,
+    ) -> Result<Self, Error> {
+        if !descriptor.has_wildcard() {
+            return Err(Error::Unexpected(
+                "Descriptors without wildcards are not supported in elip151".into(),
+            ));
+        }
+
+        // Handle multi-path
+        let script_pubkeys: Vec<_> = descriptor
+            .clone()
+            .into_single_descriptors()
+            .expect("valid descriptor")
+            .iter()
+            .map(|descriptor| {
+                // Remove wildcards
+                descriptor
+                    .at_derivation_index((1 << 31) - 1)
+                    .expect("index not hardened, not multi-path")
+                    .script_pubkey()
+            })
+            .collect();
+
+        let mut eng = Elip151Hash::engine();
+        for script_pubkey in script_pubkeys {
+            Builder::new()
+                .push_opcode(opcodes::all::OP_INVALIDOPCODE)
+                .into_script()
+                .consensus_encode(&mut eng)
+                .expect("engines don't error");
+            script_pubkey
+                .consensus_encode(&mut eng)
+                .expect("engines don't error");
+        }
+        let hash_bytes = Elip151Hash::from_engine(eng).to_byte_array();
+
+        // This computes mod n
+        let scalar = secp256k1::scalar::Scalar::from_be_bytes(hash_bytes).expect("bytes from hash");
+        let secret_key =
+            secp256k1::SecretKey::from_slice(&scalar.to_be_bytes()).expect("bytes from scalar");
+
+        // Single view keys are displayed as hex (not WIF) so we can choose any netowrk here
+        let network = Network::Bitcoin;
+        Ok(Key::View(DescriptorSecretKey::Single(SinglePriv {
+            origin: None,
+            key: bitcoin::key::PrivateKey::new(secret_key, network),
+        })))
```

Comment:
In 34005bca0e7109f3a4047e576632d694e269a8e2:

Can you remind me why we would've had trouble parsing WIF with a network here? (I don't think we should support WIF anyway for various reasons, but I'm curious now.)

Otherwise ACK.

Reply:
The main issue is converting `ct(elip151,desc)` to `ct(key,desc)` where `key` is the actual view key.
If we use WIF, we should deduce the network from `desc` which however it might not be possible (eg if `desc` has only single public keys).
There are other options to handle the conversion, but we decided to go for the "hex" solution.
These were discussed here https://github.com/ElementsProject/elements-miniscript/pull/61#issuecomment-1823943673

Other issues with WIF are that we might have mismatching network between the descriptor blinding key and the ordinary descriptor, and that single pub keys have no network, while single priv keys in WIF have a network so it's somehow inconsistent.

Reply:
Ok, the conversion thing makes sense, thanks!

The other issues, about mismatching networks etc., we have to deal with anyway with xprivs and xpubs. I think that most software "deals with it" by just ignoring the network bytes :)